### DISCUSSION TITLE
Dynamic Table creation and mapping

### DISCUSSION THREAD

**Discussion by AsfanUlla** [piccolo-orm/piccolo#904] [Q&A] [answered]
Hi i am creating a Multi-tenant system where each tenant would have their own separate tables, I was wondering is it possible to create multiple tables using the same schema and map them to the same object. here is a example on what i am trying to achieve in SqlAlchemy [https://medium.com/@ketansomvanshi007/dynamic-object-mapping-in-sqlalchemy-for-multitenancy-4e7c5064c816](https://medium.com/@ketansomvanshi007/dynamic-object-mapping-in-sqlalchemy-for-multitenancy-4e7c5064c816)

**Comment by dantownsend** [comment, marked as answer]
Piccolo lets you specify which schema a table lives in:

```python
class MyTable(Table, schema='schema_1'):
    my_column = Varchar()
```

If you potentially have many schemas (one for each customer) you can create table classes at run time using this:

https://github.com/piccolo-orm/piccolo/blob/5f972d0a7af45aeb3b18fc02f61b8c913f03b118/piccolo/table.py#L1358-L1387

For example:

```python
from piccolo.table import create_table_class
from piccolo.columns import Varchar

MyTable = create_table_class(
    class_name='MyTable',
    class_kwargs={
        "schema": 'schema_1',
    },
    class_members={
        'name': Varchar()
    },
)
```

You can then use that table like any other. For example:

```python
def get_table(schema: str):
    return create_table_class(
        class_name='MyTable',
        class_kwargs={
            "schema": schema,
        },
        class_members={
            'name': Varchar()
        },
    )

async def get_user_data(schema: str):
    MyTable = get_table(schema)
    return await MyTable.objects()
```

Another solution is to use Piccolo's [reflection API](https://piccolo-orm.readthedocs.io/en/latest/piccolo/schema/advanced.html#reflection).
